DOCSP-41161: Builders landing page (#28)
mongodb/docs-kotlin-sync: source/builders.txt

.. _kotlin-sync-builders:

================
Use Builders API

> ================

Reviewer: S: On second thought, this might be a bit clearer if we avoid the tech terms. What about

> Implement Builders Code Pattern

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. .. toctree::

..    /fundamentals/builders/aggregates
..    /fundamentals/builders/filters
..    /fundamentals/builders/indexes
..    /fundamentals/builders/projections
..    /fundamentals/builders/sort
..    /fundamentals/builders/updates

Overview
--------

> This page describes how to use the builders API and demonstrates the utility provided by
> builder classes.

Reviewer: ```suggestion
This page describes how to implement the builders pattern in your code
and describes benefits of using builder classes.
```

> The {+driver-short+} provides type-safe classes and methods that enable developers to
> efficiently build queries and aggregations.

Reviewer: ```suggestion
The {+driver-short+} provides type-safe builders classes and methods that enable developers to
efficiently build queries and aggregations. To learn more about the builders pattern, see `Type-safe builders <https://kotlinlang.org/docs/type-safe-builders.html>`__ in the kotlin docs
```
Use the source constants ^ for kotlin and the link

Author: Great idea to link out here, much obliged.

> Why Use Builders?
> -----------------
> 
> When using the MongoDB shell or plain Kotlin to construct BSON query documents, you are not
> able to identify errors in your code until runtime.
> 
> With builder classes, you can write operators as methods and any syntax errors will be
> caught at compile time.

Reviewer: ```suggestion
Why Use the Builders Code Pattern?
----------------------------------

If you use only plain {+language+} to construct BSON query documents, you are not
able to identify errors in your code until runtime.

When you use builder classes, you can write operators as methods and any syntax errors will be
caught at compile time.
```

Example
-------

> This section describes three methods of fetching the ``email`` field values of documents
> in the ``users`` collection that meet the following criteria:

Reviewer: ```suggestion
This section provides three equivalent ways to fetch the ``email`` field values of documents
in the ``users`` collection that meet the following criteria:
```

> - Users that identify as ``female``
> - Users that are older than ``29``

Reviewer: S: describe by using fields and values instead of personificaiton
```suggestion
- ``gender`` value is ``female``
- ``age`` value is greater than ``29``
```

The following data class models the documents in the ``users`` collection:

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-user-class
   :end-before: end-user-class
   :language: kotlin
   :copyable:
   :dedent:

The following data class models the results returned by our query:

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-result-class
   :end-before: end-result-class
   :language: kotlin
   :copyable:
   :dedent:

> Using the MongoDB Shell

Reviewer: MongoDB Query API

~~~~~~~~~~~~~~~~~~~~~~~

> The following sample executes the previously defined query using the MongoDB shell:

Reviewer: ```suggestion
The following sample performs the query by using the MongoDB Query API:
```

Author: Can also change the section head here to reflect

.. code-block:: js

   collection.find({ "gender": "female", "age" : { "$gt": 29 }}, { "_id": 0, "email": 1 })

> Without Using Builders
> ~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Document Class Filter
~~~~~~~~~~~~~~~~~~~~~~
```

> The following example executes the previously defined query using plain {+language+}:

Reviewer: ```suggestion
The following example performs the query by using the ``Document`` class to construct the filter:
```

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-find
   :end-before: end-find
   :language: kotlin
   :copyable:
   :dedent:

> Using Builders

Reviewer: Builders Pattern

~~~~~~~~~~~~~~

> The following example executes the previously defined query using the builders API:

Reviewer: ```suggestion
The following example performs the query by using the builders pattern:
```

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-find-builders
   :end-before: end-find-builders
   :language: kotlin
   :copyable:
   :dedent:

.. TODO: Uncomment as pages get built

.. Available Builders
.. ------------------

> 

Reviewer: S: add description

The following pages describe how to implement the different classes of builders available in the {+driver-short+}:

.. _kotlin-sync-builders:

=========================
Use Builders Code Pattern
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. .. toctree::

..    /fundamentals/builders/aggregates
..    /fundamentals/builders/filters
..    /fundamentals/builders/indexes
..    /fundamentals/builders/projections
..    /fundamentals/builders/sort
..    /fundamentals/builders/updates

Overview
--------

> This page describes how to the various available builders in your code and describes

Reviewer 2: ```suggestion
This page describes how to used the various available builders in your code and describes
```

.. _kotlin-sync-builders:

===============================
Implement Builders Code Pattern
===============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. .. toctree::

..    /fundamentals/builders/aggregates
..    /fundamentals/builders/filters
..    /fundamentals/builders/indexes
..    /fundamentals/builders/projections
..    /fundamentals/builders/sort
..    /fundamentals/builders/updates

Overview
--------

This page describes how to implement the builders pattern in your code and describes
benefits of using builder classes.

> The {+driver-short+} provides type-safe builders classes and methods that enable developers to
> efficiently build queries and aggregations. To learn more about the builders pattern,
> see `Type-safe builders <https://kotlinlang.org/docs/type-safe-builders.html>`__ in the
> {+language+} documentation.

Reviewer: S: possible convention - where we use "builder classes" use singular builder. Where we refer to builders alone, use plural
```suggestion
The {+driver-short+} provides type-safe builder classes and methods that enable developers to
efficiently build queries and aggregations. To learn more about the builders pattern,
see `Type-safe builders <https://kotlinlang.org/docs/type-safe-builders.html>`__ in the
{+language+} documentation.
```

Why Use the Builders Code Pattern?
----------------------------------

If you use only plain {+language+} to construct BSON query documents, you are not
able to identify errors in your code until runtime.

When you use builder classes, you can write operators as methods and any syntax errors

> will be caught at compile time.

Reviewer: S: I think this can be one paragraph

Example
-------

This section provides three equivalent ways to fetch the ``email`` field values of documents
in the ``users`` collection that meet the following criteria:

- ``gender`` value is ``"female"``
- ``age`` value is greater than ``29``

The following data class models the documents in the ``users`` collection:

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-user-class
   :end-before: end-user-class
   :language: kotlin
   :copyable:
   :dedent:

The following data class models the results returned by our query:

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-result-class
   :end-before: end-result-class
   :language: kotlin
   :copyable:
   :dedent:

MongoDB Query API
~~~~~~~~~~~~~~~~~

The following sample performs the query by using the MongoDB Query API:

.. code-block:: js

>    collection.find({ "gender": "female", "age" : { "$gt": 29 }}, { "_id": 0, "email": 1 })

Reviewer: S: reduce horiz scroll
```suggestion
   collection.find(
       { "gender": "female", "age" : { "$gt": 29 }},
       { "_id": 0, "email": 1 }
   )
```

.. _kotlin-sync-builders:

===============================

> Implement Builders Code Pattern

Reviewer 2: Not sure this title makes sense - users aren't implementing builders we are consuming them.

===============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. .. toctree::

..    /fundamentals/builders/aggregates
..    /fundamentals/builders/filters
..    /fundamentals/builders/indexes
..    /fundamentals/builders/projections
..    /fundamentals/builders/sort
..    /fundamentals/builders/updates

Overview
--------

> This page describes how to implement the builders pattern in your code and describes

Reviewer 2: Nit: This seems a little verbose. In short its about using builders to help build operations in a less verbose way than using Documents.

Author: For further clarification, do you think the entire Overview section is too verbose?

Reviewer 2: Will add suggesions

> This page describes how to implement the builders pattern in your code and describes
> benefits of using builder classes.
> 
> The {+driver-short+} provides type-safe builder classes and methods that enable developers to
> efficiently build queries and aggregations. To learn more about the builders pattern,
> see `Type-safe builders <https://kotlinlang.org/docs/type-safe-builders.html>`__ in the
> {+language+} documentation.

Reviewer 2: ```suggestion
This page describes how to use the various available builders in your code and describes
benefits of using the provided builders.

The {+driver-short+} provides type-safe builder classes and methods that enable developers to
efficiently build queries and aggregations.
```

Reviewer 2: Note: https://kotlinlang.org/docs/type-safe-builders.html may not be appropiate here, as the builders used are written in Java and don't/can't catch the type information in the same way as the examples in the kotlin documentation.

> Why Use the Builders Code Pattern?

Reviewer 2: ```suggestion
Why Use Builders?
```

----------------------------------

> If you use only plain {+language+} to construct BSON query documents, you are not
> able to identify errors in your code until runtime. When you use builder classes, you can
> write operators as methods and any syntax errors will be caught at compile time.

Reviewer 2: ```suggestion
If you use only plain {+language+} to construct BSON query documents, you are not
able to identify syntax errors until runtime. The builders help ensure the correctness of syntax and can be less verbose than constructing BSON documents.
```

in the ``users`` collection that meet the following criteria:

- ``gender`` value is ``"female"``
- ``age`` value is greater than ``29``

The following data class models the documents in the ``users`` collection:

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-user-class
   :end-before: end-user-class
   :language: kotlin
   :copyable:
   :dedent:

The following data class models the results returned by our query:

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-result-class
   :end-before: end-result-class
   :language: kotlin
   :copyable:
   :dedent:

MongoDB Query API
~~~~~~~~~~~~~~~~~

The following sample performs the query by using the MongoDB Query API:

.. code-block:: js

   collection.find(
       { "gender": "female", "age" : { "$gt": 29 }},
       { "_id": 0, "email": 1 }
   )

Document Class Filter
~~~~~~~~~~~~~~~~~~~~~

The following example performs the query by using the ``Document`` class to construct the
query filter:

.. literalinclude:: /includes/builders/builders.kt
   :start-after: start-find
   :end-before: end-find
   :language: kotlin
   :copyable:
   :dedent:

Builders Pattern
~~~~~~~~~~~~~~~~

> The following example performs the query by using the builders pattern:

Reviewer 2: ```suggestion
The following example performs the query by using the builder helpers:
```
